File: stripe/src/test/java/com/stripe/android/PaymentAuthWebViewStarterTest.java
package com.stripe.android;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import com.stripe.android.stripe3ds2.init.ui.StripeToolbarCustomization;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class PaymentAuthWebViewStarterTest {
    private static final String CLIENT_SECRET =
            "pi_1EceMnCRMbs6FrXfCXdF8dnx_secret_vew0L3IGaO0x9o0eyRMGzKr0k";
    private static final PaymentAuthWebViewStarter.Data DATA =
            new PaymentAuthWebViewStarter.Data(
                    CLIENT_SECRET,
                    "https://hooks.stripe.com/",
                    "stripe://payment-auth"
            );

    @Mock private Activity mActivity;
    @Captor private ArgumentCaptor<Intent> mIntentArgumentCaptor;
    @Captor private ArgumentCaptor<Integer> mRequestCodeCaptor;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void start_startsWithCorrectIntentAndRequestCode() {
        new PaymentAuthWebViewStarter(mActivity, 50000).start(DATA);
        verify(mActivity).startActivityForResult(mIntentArgumentCaptor.capture(),
                mRequestCodeCaptor.capture());

        final Intent intent = mIntentArgumentCaptor.getValue();
        final Bundle extras = intent.getExtras();
        assertNotNull(extras);
        assertNull(extras.getParcelable(PaymentAuthWebViewStarter.EXTRA_UI_CUSTOMIZATION));
        assertEquals(4, extras.size());
        assertEquals(CLIENT_SECRET,
                extras.getString(PaymentAuthWebViewStarter.EXTRA_CLIENT_SECRET));
    }

    @Test
    public void start_startsWithCorrectIntentAndRequestCodeAndCustomization() {
        new PaymentAuthWebViewStarter(mActivity, 50000,
                new StripeToolbarCustomization()).start(DATA);
        verify(mActivity).startActivityForResult(mIntentArgumentCaptor.capture(),
                mRequestCodeCaptor.capture());

        final Intent intent = mIntentArgumentCaptor.getValue();
        final Bundle extras = intent.getExtras();
        assertNotNull(extras);
        assertNotNull(extras.getParcelable(PaymentAuthWebViewStarter.EXTRA_UI_CUSTOMIZATION));
        assertEquals(4, extras.size());
        assertEquals(CLIENT_SECRET,
                extras.getString(PaymentAuthWebViewStarter.EXTRA_CLIENT_SECRET));
    }
}

File: stripe/src/test/java/com/stripe/android/PaymentAuthWebViewStarterTest.kt
package com.stripe.android

import android.app.Activity
import android.content.Intent
import androidx.test.core.app.ApplicationProvider
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.stripe.android.auth.PaymentAuthWebViewContract
import com.stripe.android.payments.DefaultReturnUrl
import com.stripe.android.stripe3ds2.init.ui.StripeToolbarCustomization
import com.stripe.android.view.AuthActivityStarter
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

@RunWith(RobolectricTestRunner::class)
class PaymentAuthWebViewStarterTest {

    private val activity: Activity = mock()

    private val intentArgumentCaptor: KArgumentCaptor<Intent> = argumentCaptor()
    private val requestCodeCaptor: KArgumentCaptor<Int> = argumentCaptor()

    private val defaultReturnUrl = DefaultReturnUrl.create(
        ApplicationProvider.getApplicationContext()
    )
    private val host = AuthActivityStarter.Host.create(activity)
    private val legacyStarter = PaymentAuthWebViewStarter.Legacy(
        host,
        isCustomTabsSupported = true,
        defaultReturnUrl
    )

    @Test
    fun start_startsWithCorrectIntentAndRequestCode() {
        legacyStarter.start(DATA)
        verify(activity).startActivityForResult(
            intentArgumentCaptor.capture(),
            requestCodeCaptor.capture()
        )

        val args = requireNotNull(
            PaymentAuthWebViewContract.parseArgs(intentArgumentCaptor.firstValue)
        )
        assertNull(args.toolbarCustomization)
        assertEquals(DATA.clientSecret, args.clientSecret)
    }

    @Test
    fun start_startsWithCorrectIntentAndRequestCodeAndCustomization() {
        legacyStarter.start(
            DATA.copy(
                toolbarCustomization = StripeToolbarCustomization()
            )
        )
        verify(activity).startActivityForResult(
            intentArgumentCaptor.capture(),
            requestCodeCaptor.capture()
        )

        val args = requireNotNull(
            PaymentAuthWebViewContract.parseArgs(intentArgumentCaptor.firstValue)
        )
        assertNotNull(args.toolbarCustomization)
        assertEquals(DATA.clientSecret, args.clientSecret)
    }

    private companion object {
        private val DATA = PaymentAuthWebViewContract.Args(
            objectId = "pi_1EceMnCRMbs6FrXfCXdF8dnx",
            requestCode = 50000,
            clientSecret = "pi_1EceMnCRMbs6FrXfCXdF8dnx_secret_vew0L3IGaO0x9o0eyRMGzKr0k",
            url = "https://hooks.stripe.com/",
            returnUrl = "stripe://payment-auth"
        )
    }
}
